File: PaleoRose/Classes/Graphics/XRGraphicCircle.m
//
//  XRGraphicCircle.m
//  XRose
//
//  Created by Tom Moore on Sat Jan 24 2004.
//
// MIT License
//
// Copyright (c) 2004 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#import "XRGraphicCircle.h"
#import "XRGeometryController.h"

@implementation XRGraphicCircle

-(instancetype)initWithController:(XRGeometryController *)controller {
    if(self = [super initWithController:controller]) {
        _isPercent = [controller isPercent];
        [self registerForKVO];
        if([self isMemberOfClass:[XRGraphicCircle class]]) // this prevents calculating geometry twice for circle labels
            [self calculateGeometry];
    }
    return self;
}

-(id)initCoreCircleWithController:(XRGeometryController *)aController {
	if(self = [super initWithController:aController])
	{
        self.countSetting = 0;
		_percentSetting = 0.0;
        [self registerForKVO];
		_isPercent = [aController isPercent];
		if([self isMemberOfClass:[XRGraphicCircle class]])
			[self calculateGeometry];
	}
	return self;
}

-(void)registerForKVO {
    int i = 0;
    NSArray *keys = @[@"countSetting", @"percentSetting"];
    for(i=0;i<keys.count;i++) {
        [self addObserver:self
               forKeyPath:keys[i]
                  options:NSKeyValueObservingOptionNew |
         NSKeyValueObservingOptionOld
                  context:NULL];
    }
}

-(void)setGeometryPercent:(float)percent {
	self.isGeometryPercent = YES;
	self.drawingPath = [NSBezierPath bezierPathWithOvalInRect:[self.geometryController circleRectForGeometryPercent:percent]];
}

-(void)calculateGeometry {
	
	if([self.geometryController isPercent])
	{
		self.drawingPath = [NSBezierPath bezierPathWithOvalInRect:[self.geometryController circleRectForPercent:_percentSetting]];
	}
	else
	{
		self.drawingPath = [NSBezierPath bezierPathWithOvalInRect:[self.geometryController circleRectForCount:_countSetting]];
	}
}

-(NSDictionary *)graphicSettings {
    NSMutableDictionary *parentDict = [NSMutableDictionary dictionaryWithDictionary:[super graphicSettings]];
    NSDictionary *classDict = @{
        XRGraphicKeyGraphicType : @"Circle",
        XRGraphicKeyCountSetting : [self stringFromInt:self.countSetting],
        XRGraphicKeyPercentSetting : [self stringFromFloat:_percentSetting],
        XRGraphicKeyGeometryPercent : [self stringFromFloat:_percentSetting],
        XRGraphicKeyIsGeometryPercent : [self stringFromBool:self.isGeometryPercent],
        XRGraphicKeyIsPercent : [self stringFromBool:_isPercent],
        XRGraphicKeyIsFixedCount : [self stringFromBool:_isFixedCount]
    };
    [parentDict addEntriesFromDictionary: classDict];
	return parentDict;
}

-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    [super observeValueForKeyPath:keyPath
                             ofObject:object
                               change:change
                              context:context];
    if ([keyPath isEqualToString:@"countSetting"]) {
        _isPercent = NO;
        self.isGeometryPercent = NO;
        [self calculateGeometry];
    } else if ([keyPath isEqualToString:@"percentSetting"]) {
        _isPercent = YES;
        self.isGeometryPercent = NO;
        [self calculateGeometry];
    }
}
@end

File: PaleoRose/Classes/Graphics/XRGraphicCircle.swift
//
//  XRGraphicCircle.swift
//  PaleoRose
//
//  Created by Migration Assistant on 2025-08-29.
//
// MIT License
//
// Copyright (c) 2004 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import AppKit

@objc class XRGraphicCircle: XRGraphic {

    // MARK: - Properties

    @objc dynamic var isFixedCount: Bool = false
    @objc dynamic var countSetting: Int32 = 0 {
        didSet {
            isPercent = false
            isGeometryPercent = false
            calculateGeometry()
        }
    }

    @objc dynamic var percentSetting: Float = 0.0 {
        didSet {
            isPercent = true
            isGeometryPercent = false
            calculateGeometry()
        }
    }

    @objc dynamic var isGeometryPercent: Bool = false
    @objc dynamic var isPercent: Bool = false

    // MARK: - KVO Keys

    private static let kvoKeyCountSetting = "countSetting"
    private static let kvoKeyPercentSetting = "percentSetting"

    // MARK: - Initialization

    @objc override init(controller: GraphicGeometrySource) {
        super.init(controller: controller)
        isPercent = controller.isPercent()
        registerForKVO()

        // Prevent calculating geometry twice for circle labels
        if type(of: self) == XRGraphicCircle.self {
            calculateGeometry()
        }
    }

    @objc(initCoreCircleWithController:) init(coreCircleWithController controller: GraphicGeometrySource) {
        super.init(controller: controller)
        countSetting = 0
        percentSetting = 0.0
        isPercent = controller.isPercent()
        registerForKVO()

        // Prevent calculating geometry twice for circle labels
        if type(of: self) == XRGraphicCircle.self {
            calculateGeometry()
        }
    }

    private func registerForKVO() {
        let keyPaths = [XRGraphicCircle.kvoKeyCountSetting, XRGraphicCircle.kvoKeyPercentSetting]
        for keyPath in keyPaths {
            addObserver(self, forKeyPath: keyPath, options: [.new, .old], context: nil)
        }
    }

    deinit {
        let keyPaths = [XRGraphicCircle.kvoKeyCountSetting, XRGraphicCircle.kvoKeyPercentSetting]
        for keyPath in keyPaths {
            removeObserver(self, forKeyPath: keyPath)
        }
    }

    // MARK: - Geometry

    @objc func setGeometryPercent(_ percent: Float) {
        isGeometryPercent = true
        guard let controller = geometryController else { return }
        let circleRect = controller.circleRect(forGeometryPercent: percent)
        drawingPath = NSBezierPath(ovalIn: circleRect)
    }

    @objc override func calculateGeometry() {
        guard let controller = geometryController else { return }

        var circleRect = NSRect.zero
        if controller.isPercent() {
            circleRect = controller.circleRect(forPercent: percentSetting)
        } else {
            circleRect = controller.circleRect(forCount: countSetting)
        }

        drawingPath = NSBezierPath(ovalIn: circleRect)
        drawingPath?.lineWidth = CGFloat(lineWidth)
    }

    // MARK: - Settings Export

    @objc override func graphicSettings() -> [AnyHashable: Any] {
        var settings = super.graphicSettings()

        settings["GraphicType"] = "Circle"
        settings["_countSetting"] = string(from: countSetting)
        settings["_percentSetting"] = string(from: percentSetting)
        settings["_geometryPercent"] = string(from: percentSetting)
        settings["_isGeometryPercent"] = string(from: isGeometryPercent)
        settings["_isPercent"] = string(from: isPercent)
        settings["_isFixedCount"] = string(from: isFixedCount)

        return settings
    }

    // MARK: - KVO

    @objc override func observeValue(forKeyPath keyPath: String?, of object: Any?,
                                     change: [NSKeyValueChangeKey: Any]?, context: UnsafeMutableRawPointer?)
    {
        super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)

        guard let keyPath else { return }

        if keyPath == XRGraphicCircle.kvoKeyCountSetting {
            isPercent = false
            isGeometryPercent = false
            calculateGeometry()
        } else if keyPath == XRGraphicCircle.kvoKeyPercentSetting {
            isPercent = true
            isGeometryPercent = false
            calculateGeometry()
        }
    }
}
